aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -2,13 +2,506 @@
+      // rarity in which this occurs, this should not be an issue.
+      for (int cat = categories - 1; cat >= 0; --cat) {
+        if (dist[curDist * stride_dist + cat * stride_categories] > zero) {
+          dest[curDist] = cat;
+          break;
+        }
+      }
+    }
+  }
+}
+
+
+void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t n_sample, const bool with_replacement, Generator* generator) {
+  auto gen = get_generator_or_default<CUDAGenerator>(generator, cuda::detail::getDefaultCUDAGenerator());
+
+  int inputSize = self.dim();
+  int64_t numDist =
+      inputSize == 1 ? 1 : self.size(0);
+  int numCategories =
+      inputSize == 1 ? self.size(0) : self.size(1);
+
+  // Restructure data for 2d
+  if (inputSize == 1) {
+    self.unsqueeze_(0);
+  }
+
+  result.resize_({numDist, n_sample});
+
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "multinomial_kernel_cuda", [&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+    auto props = at::cuda::getCurrentDeviceProperties();
+    assert(props != NULL);
+    int numSM = props->multiProcessorCount;
+    int maxThreads = props->maxThreadsPerBlock;
+    int maxShared = props->sharedMemPerBlock;
+    int requiredShared = (numCategories < maxThreads ? numCategories : maxThreads)
+                         * (sizeof(scalar_t) + sizeof(accscalar_t));
+
+    if (n_sample == 1 && maxShared >= requiredShared) {
+      // Optimized allocation-free implementation
+      // To exploit greater parallelism for the sampling, generate the
+      // Uniform random samples in a separate kernel launch, into
+      // temporarily allocated memory. The device RNG is thread-limited
+      Tensor sampled = at::empty({numDist, n_sample}, self.options());
+      at::native::uniform_cuda_(sampled, 0.0, 1.0, gen);
+
+      dim3 block(numCategories < maxThreads ? numCategories : maxThreads);
+      dim3 grid(numDist < numSM * 4 ? numDist : numSM * 4);
+
+      sampleMultinomialOnce<scalar_t, accscalar_t>
+          <<<grid, block,
+          requiredShared,
+          at::cuda::getCurrentCUDAStream()>>>(
+              result.data_ptr<int64_t>(),
+                  numDist,
+                  numCategories,
+                  sampled.data_ptr<scalar_t>(),
+                  self.data_ptr<scalar_t>(),
+                  self.stride(0),
+                  self.stride(1)
+          );
+    } else {
+      // Generic, slow implementation with memory allocations
+
+      // For sampling without replacement, we modify the distribution
+      // for subsequent samples in this space
+      Tensor origDist = at::empty_like(self);
+      origDist.copy_(self);
+
+      Tensor normDist = at::empty_like(self);
+
+      Tensor prefixSum = at::empty_like(self);
+
+      // Renorm along rows
+      normDist.copy_(origDist);
+      renormRows(normDist);
+
+      // Prefix sum along rows
+      legacy::cuda::_th_cumsum_out(prefixSum, normDist, 1);
```

Comment:
Can we call it via at:: dispatch?

Reply:
I think in native/cuda we use legacy::cuda::_th_xxx in most cases. Why should we do this via dispatch? I think in future, when we move _th_cumsum_out to ATen, it should be changed to direct call.